patroni/postgresql/config.py

```diff
@@ -271,6 +271,31 @@ def _bool_is_true_validator(value: Any) -> bool:
     return parse_bool(value) is True
 
 
+class ParamDiff(Dict[str, str]):
+    """Single PG parameter's value diff to be stored in :attr:`~patroni.postgresql.Postgresql._pending_restart_reason`.
+
+    Extends :class:`dict` to only have the two specific attributes: ``old_value`` and ``new_value`` and their values
+    casted to :class:`str` and converted from base units (if possible).
+    """
+
+    def __init__(self, old_value: Any, new_value: Any, vartype: Optional[str] = None, unit: Optional[str] = None):
+        """Create a new instance of :class:`ParamDiff` with the given *old_value* and *new_value*.
+
+        :param old_value: current :class:`str` parameter value.
+        :param new_value: :class:`str` value of the paramater after a restart.
+        :param vartype: the target type to parse old/new_value. See ``vartype`` argument of
+            :func:`~patroni.utils.maybe_convert_from_base_unit`.
+        :param unit: unit of *old/new_value*. See ``base_unit`` argument of
+            :func:`~patroni.utils.maybe_convert_from_base_unit`.
+        """
+        super().__init__()
+        str_value: Callable[[Any], str] = lambda x: '' if x is None else str(x)
+        self['old_value'] = (maybe_convert_from_base_unit(str_value(old_value), vartype, unit)
+                             if vartype else str_value(old_value))
+        self['new_value'] = (maybe_convert_from_base_unit(str_value(new_value), vartype, unit)
+                             if vartype else str_value(new_value))
```

Comment:
It seems to be overengineered.
Instead it could be just a function, that returns a dict with two key-value pairs.